File: Pascal/Task600.pas
program Task600;
uses sysutils;
{Оголошуємо тип даних для опису однієї книги}
type book = record
    publisher : string[50];
    book_title : string[255];
    author : string[100];
    publish_year : integer;
    price : real;
    count : integer
    end;
const db_name : string = 'books_shop.db';
var tmp : book;
    database : file of book;
    choise : integer;
    
procedure readall(filename: string);
    var data: book;
begin
    assign(database, filename);
    reset(database);
    
    while eof(database)=false do
    begin
        read(database,data);
        writeln(data.author,'  ',data.book_title,
        ' ', data.publisher,' ',data.publish_year);
    end;
    close(database);
end;

procedure find_by_author(filename: string);
    var data: book;
    author : string[100]; {змінна для пошуку 
    співпадає за типом з відповідним полем запису}
begin
    assign(database, filename);
    reset(database);
    write('Введіть прізвище автора:');
    readln(author);
    while eof(database)=false do
    begin
        read(database,data);
        if pos(author, data.author) > 0 then
            writeln(data.author,'  ',data.book_title,
            ' ', data.publisher,' ',data.publish_year);
    end;
    close(database);
end;

procedure find_book(filename: string);
    var data: book;
    book_title : string[255];{локальна змінна для
        назви книги, або фрагмента назви}
begin
    assign(database, filename);
    reset(database);
    write('Введіть назву книги, або її частину:');
    readln(book_title);
    while eof(database)=false do
    begin
        read(database,data);
        if pos(book_title, data.book_title) > 0 then
            writeln(data.author,'  ',data.book_title,
            ' ', data.publisher,' ',data.publish_year);
    end;
    close(database);
end;

procedure read_by_publisher(filename: string);
    var data: book;
    publisher : string[50]; {локальна змінна для пошуку 
    співпадає за типом з відповідним полем запису}
begin
    assign(database, filename);
    reset(database);
    write('Видавництво:'); readln(publisher);
    while eof(database) = false do
    begin
        read(database,data);
        if data.publisher = publisher then
        writeln(data.author,'  ',data.book_title,
        ' ', data.publisher,' ',data.publish_year);
    end;
    close(database);
end;

function inputbook() : book;
    var result : book;
begin
    writeln('Введіть про нову книгу');
    write('Автор(автори):'); readln(result.author);
    write('Назва:'); readln(result.book_title);
    write('Видавництво:'); readln(result.publisher);
    write('Рік видання:'); readln(result.publish_year);
    write('Ціна:'); readln(result.price);
    write('Кількість:'); readln(result.count);
    inputbook := result;
end;

procedure writenewfile(filename: string);
var choise: integer; 
begin
    assign(database, filename);
    rewrite(database);
    repeat
    tmp := inputbook();  
    write(database, tmp);
    writeln('Додати ще одну книгу?');
    writeln('1 - так, додати ще;');
    writeln('0 - зберегти дані і вийти');
    readln(choise);
    until(choise<>1);
    writeln('нову базу книг стоврено');
    close(database);
end;

procedure writebook(filename: string; data: book);
const tmp_filename = 'database.dump';
var dump_file: file of book;
begin
    assign(database, filename);
    reset(database);
    assign(dump_file, tmp_filename);
    rewrite(dump_file);
    while eof(database)=false do
    begin
        read(database,tmp);
        write(dump_file,tmp)
    end;
    
    write(dump_file, data);
    close(dump_file);
    rewrite(database);
    reset(dump_file);
    while eof(dump_file)=false do
    begin
        read(dump_file,tmp);
        write(database,tmp)
    end;
    writeln('книга додана в базу');
    close(database);
    close(dump_file);
    deletefile(tmp_filename)
end;

begin
  repeat
    writeln('Оберіть потрібну дію:');
    writeln('1 - шукати книгу за назвою.');
    writeln('2 - шукати книги за прізвищем автора.');
    writeln('3 - вивести книги вказаного видавництва');
    writeln('4 - переглянути всі книги в базі');
    writeln('5 - створити/перестворити базу книг.');
    writeln('6 - дописати нову книгу в базу');
    writeln('0 - завершити роботу');
    readln(choise);
    case (choise) of
      1:find_book(db_name);
      2:find_by_author(db_name);
      3:read_by_publisher(db_name);
      4:readall(db_name);
      5:writenewfile(db_name);
      6:writebook(db_name, inputbook());
    end;
  until(choise = 0)
end.

File: C++/Task600.cpp
#include <iostream>
#include <fstream>
#include "Windows.h"
using namespace std;
//Оголошуємо тип даних для опису однієї книги
struct book
{
  char publisher[50];
  char book_title[255];
  char author[100];
  int publish_year;
  double price;
  int count;
};
//назва файлу з даними
const char db_name[] = "books_shop.db";
//прототип функції для читання усіх книг
void readall(const char *);
//прототип функції для вводу даних про окрему книгу
book inputbook();
//функція для створення нового файлу з книгами 
void writenewfile(const char* filename);
//прототип функції для дозапису нової книги у файл
void writebook(const char*, book);
//прототип функції для пошуку книг за авторами
void show_by_authors(const char*, const char*);
//прототип функції для пошуку книг за назвою
void show_by_name(const char*, const char*);
//прототип функції для пошуку книг за видавництвом
void show_by_publisher(const char*, const char*);

//головна функція (програма)
int main()
{
  SetConsoleCP(1251); SetConsoleOutputCP(1251);
  int choise;
  do {
    system("cls");
    cout << "Оберіть потрібну дію : \n";
    cout << "1 - шукати книги за прізвищем автора \n";
    cout << "2 - шукати книги за назвою \n";
    cout << "3 - шукати книги за видавництвом \n";
    cout << "4 - переглянути всі книги в базі \n";
    cout << "5 - створити/перестворити базу книг. \n";
    cout << "6 - дописати нову книгу в базу \n";
    cout << "0 - завершити роботу програми \n";
    cin >> choise; cin.get();
    char buff[255] = {};//фрагмент тексту для пошуку
    switch (choise) {
    case 1:
      cout << "Введіть прізвище автора: ";
      cin.getline(buff, 255);
      show_by_authors(db_name, buff);
      break;
    case 2:
      cout << "Введіть назву книги або її частину: ";
      cin.getline(buff, 255);
      show_by_name(db_name, buff);
      break;
    case 3:
      cout << "Введіть назву видавництва: ";
      cin.getline(buff, 255);
      show_by_publisher(db_name, buff);
      break;
    case 4:
      readall(db_name);
      break;
    case 5:
      writenewfile(db_name);
      break;
    case 6:
      book tmp = inputbook();
      writebook(db_name, tmp);
      break;
    }
  } while (choise != 0);
  
  return 0;
}

void readall(const char* filename)
{
  book data;
  FILE* database;
  fopen_s(&database, filename, "rb");
  while (true)
  {
    int count = fread(&data, sizeof(data), 1, database);
    if (count == 0) break;
    cout << data.author << " " << data.book_title
      << " " << data.publisher << " " 
        << data.publish_year << endl;
  }
  system("pause");
  fclose(database);
}
 
book inputbook()
{
  book result;
  cout << "Введіть інформацію про нову книгу\n";
  cout << "Автор(автори):";
  cin.getline(result.author, 100);
/* функція getline використовується для читання з консолі
    тексту влючно з пробілами та розділовими знаками,
    результат читання поміщається в змінну, передану першим
    параметром, а другий її параметр обмежує максимальну 
    кількість символів, яку можна прочитати*/
  cout << "Назва:";
  cin.getline(result.book_title, 255);
  cout << "Видавництво:";
  cin.getline(result.publisher, 50);
  cout << "Рік видання:"; cin >> result.publish_year;
  cout << "Ціна:"; cin >> result.price;
  cout << "Кількість:"; cin >> result.count;
  cin.get();//зчитує символ закінчення рядка
  //який залишається після вводу числових даних
  //і може бути сприйнятий програмою як текст
  //при наступному вводі текстових даних
  return result;
}

void writenewfile(const char* filename)
{
  FILE* database;
  int choise;
  book tmp;
  fopen_s(&database, filename, "wb");
  do {
    tmp = inputbook();
    fwrite(&tmp, sizeof(tmp), 1, database);
    cout << "Додати ще одну книгу ? \n";
    cout << "1 - так, додати ще;\n";
    cout << "0 - зберегти дані і вийти.\n";
    cin >> choise; cin.get();
  }while(choise == 1);
  cout << "нову базу книг стоврено"<< endl;
  fclose(database);
}

void writebook(const char* filename, book data)
{
  FILE *database, *dump_file;
  // відкриваємо існуючий файл "бази даних" 
  // для запису даних в кінець файлу
  fopen_s(&database, filename, "ab");
  //дописуємо нову книгу в файл та закриваємо його
  fwrite(&data, sizeof(book), 1, database);
  // записуємо дані з буфера на диск
  fflush(database);
  // закриваємо файл
  fclose(database);
  cout << "книга додана в базу" << endl;
  system("pause");
}

void show_by_authors(const char* filename, 
                               const char* author)
{
  book data;
  FILE* database;
  fopen_s(&database, filename, "rb");
  while (true)
  {
    int count = fread(&data, sizeof(data), 1, database);
    if (count == 0) break;
    if (strstr(data.author, author) != NULL)
    {/*Функція strstr повертає покажчик на перший символ 
     першого входження вказаного фрагмента 
     у задану стрічку. Якщо співпадіння не 
     знайдено, - повертається нульовий вказівник NULL*/
      cout << data.author << " " << data.book_title
        << " " << data.publisher << " "
        << data.publish_year << endl;
    }
  }
  system("pause");
}

void show_by_name(const char* filename, 
                            const char* title_fragment)
{
  book data;
  FILE* database;
  fopen_s(&database, filename, "rb");
  while (true)
  {
    int count = fread(&data, sizeof(data), 1, database);
    if (count == 0) break;
    if (strstr(data.book_title, title_fragment) != NULL)
    {
      cout << data.author << " " << data.book_title
        << " " << data.publisher << " "
        << data.publish_year << endl;
    }
  }
  system("pause");

}

void show_by_publisher(const char* filename, 
                                const char* publisher)
{
  book data;
  FILE* database;
  fopen_s(&database, filename, "rb");
  while (true)
  {
    int count = fread(&data, sizeof(data), 1, database);
    if (count == 0) break;
    if (strcmp(data.publisher, publisher) == 0)
    {
      cout << data.author << " " << data.book_title
        << " " << data.publisher << " "
        << data.publish_year << endl;
    }
  }
  system("pause");
}
